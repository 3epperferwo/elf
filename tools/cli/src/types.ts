export interface Options {
  storeName: string;
  features: Array<Features>;
  crud: Array<'addEntities' | 'updateEntities' | 'deleteEntities'>;
  idKey: string;
  path: string;
}

export type Features = typeof baseFeatures[number]['value'];

export const baseFeatures = [
  { name: 'Props', value: 'withProps' },
  { name: 'Entities', value: 'withEntities' },
  {
    name: 'UIEntities',
    value: 'withUIEntities',
  },
  {
    name: 'Active Id',
    value: 'withActiveId',
  },
  {
    name: 'Active Ids',
    value: 'withActiveIds',
  },
  {
    name: 'Cache Status',
    value: 'withCacheStatus',
  },
  {
    name: 'Status',
    value: 'withStatus',
  },
] as const;
