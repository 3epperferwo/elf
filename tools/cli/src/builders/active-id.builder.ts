import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { camelize, capitalize } from '../utils';
import { factory } from 'typescript';

export class ActiveIdBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withActiveId';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveId'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(['withActiveId', 'selectActiveEntity', 'setActiveId']);

    this.repo.insertMember(0, {
      name: `active${capitalize(camelize(this.singularName))}$`,
      kind: StructureKind.Property,
      initializer: `store.pipe(selectActiveEntity())`,
    });

    this.repo.addMember({
      kind: StructureKind.Method,
      name: `setActiveId`,
      parameters: [
        {
          name: 'id',
          type: `${capitalize(this.singularName)}['${this.idKey}']`,
        },
      ],
      statements: [`store.reduce(setActiveId(id));`],
    });
  }
}
