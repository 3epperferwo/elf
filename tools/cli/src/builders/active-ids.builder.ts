import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { camelize, capitalize } from '../utils';
import { factory } from 'typescript';

export class ActiveIdsBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withActiveIds';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveIds'),
      undefined,
      []
    );
  }

  run() {
    this.addNamedImport([
      'withActiveIds',
      'selectActiveEntities',
      'toggleActiveIds',
    ]);

    this.repo.insertMember(0, {
      name: `active${capitalize(camelize(this.storeName))}$`,
      kind: StructureKind.Property,
      initializer: `store.pipe(selectActiveEntities())`,
    });

    this.repo.addMember({
      kind: StructureKind.Method,
      name: `toggleActiveIds`,
      parameters: [
        {
          name: 'ids',
          type: `Array<${capitalize(this.singularName)}['${this.idKey}']>`,
        },
      ],
      statements: [`store.reduce(toggleActiveIds(ids));`],
    });
  }
}
