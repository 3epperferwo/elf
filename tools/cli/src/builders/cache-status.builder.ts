import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { factory } from 'typescript';

export class CacheStatusBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withCacheStatus';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withCacheState'),
      undefined,
      []
    );
  }

  run() {
    this.addNamedImport([
      'withCacheStatus',
      'selectCacheStatus',
      'setCacheStatus',
      'CacheStatus',
    ]);

    this.repo.insertMember(0, {
      name: `cacheStatus$`,
      kind: StructureKind.Property,
      initializer: `store.pipe(selectCacheStatus())`,
    });

    this.repo.addMember({
      name: `updateCacheStatus`,
      kind: StructureKind.Method,
      parameters: [{ name: 'value', type: 'CacheStatus' }],
      statements: `store.reduce(setCacheStatus(value));`,
    });
  }
}
