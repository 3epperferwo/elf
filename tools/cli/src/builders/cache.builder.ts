import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { factory } from 'typescript';

export class CacheBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withCache';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withCache'),
      undefined,
      []
    );
  }

  run() {
    this.addNamedImport(['withCache', 'updateCache', 'CacheStatus']);

    this.repo.addMember({
      name: `updateCache`,
      kind: StructureKind.Method,
      parameters: [{ name: 'value', type: 'CacheStatus' }],
      statements: `store.reduce(updateCache(value));`,
    });
  }
}
