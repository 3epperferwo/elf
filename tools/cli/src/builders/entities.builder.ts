import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { camelize, capitalize } from '../utils';
import { factory } from 'typescript';

export class EntitiesBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withEntities';
  }

  getPropsFactory() {
    const type: any[] = [
      factory.createTypeReferenceNode(
        factory.createIdentifier(capitalize(this.singularName)),
        undefined
      ),
    ];

    if (this.idKey !== 'id') {
      type.push(
        factory.createLiteralTypeNode(factory.createStringLiteral('_id', true))
      );
    }

    let props: any[] = [];

    if (this.idKey !== 'id') {
      props = [
        factory.createObjectLiteralExpression(
          [
            factory.createPropertyAssignment(
              factory.createIdentifier('idKey'),
              factory.createStringLiteral('_id', true)
            ),
          ],
          false
        ),
      ];
    }

    return factory.createCallExpression(
      factory.createIdentifier('withEntities'),
      type,
      props
    );
  }

  run() {
    this.addNamedImport(['withEntities', 'selectAll', ...this.options.crud]);

    this.sourceFile.insertInterface(1, {
      name: capitalize(this.singularName),
      isExported: true,
      properties: [
        {
          name: this.idKey,
          type: 'number',
        },
      ],
    });

    this.repo.addMember({
      name: `${camelize(this.storeName)}$`,
      kind: StructureKind.Property,
      initializer: `store.pipe(selectAll())`,
    });

    this.options.crud.forEach((op) => this[op]?.());
  }

  addEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `add${capitalize(this.singularName)}`,
      parameters: [
        {
          name: camelize(this.singularName),
          type: capitalize(this.singularName),
        },
      ],
      statements: [
        `store.reduce(addEntities(${camelize(this.singularName)}));`,
      ],
    });
  }

  updateEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `update${capitalize(this.singularName)}`,
      parameters: [
        {
          name: 'id',
          type: `${capitalize(this.singularName)}['${this.idKey}']`,
        },
        {
          name: camelize(this.singularName),
          type: `Partial<${capitalize(this.singularName)}>`,
        },
      ],
      statements: [
        `store.reduce(updateEntities(id, ${camelize(this.singularName)}));`,
      ],
    });
  }

  removeEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `delete${capitalize(this.singularName)}`,
      parameters: [
        {
          name: 'id',
          type: `${capitalize(this.singularName)}['${this.idKey}']`,
        },
      ],
      statements: [`store.reduce(removeEntities(id));`],
    });
  }
}
