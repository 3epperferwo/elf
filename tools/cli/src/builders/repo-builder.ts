import {
  ClassDeclaration,
  printNode,
  Project,
  QuoteKind,
  SourceFile,
  StructureKind,
  VariableDeclarationKind,
} from 'ts-morph';
import { CallExpression, factory, ScriptTarget } from 'typescript';
import { RequestsCacheBuilder } from './requests-cache.builder';
import { ActiveIdsBuilder } from './active-ids.builder';
import { EntitiesBuilder } from './entities.builder';
import { UIEntitiesBuilder } from './ui-entities.builder';
import { RequestsStatusBuilder } from './requests-status.builder';
import { ActiveIdBuilder } from './active-id.builder';
import { PropsBuilder } from './props.builder';
import { camelize, capitalize } from '../utils';
import { Options } from '../types';

export function createRepo(options: Options) {
  const { storeName } = options;

  const project = new Project({
    manipulationSettings: {
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      target: ScriptTarget.ES2015,
    },
  });

  const sourceFile = project.createSourceFile(`repo.ts`, ``);

  const repoClassDec = sourceFile.addClass({
    name: `${capitalize(storeName)}Repository`,
    isExported: true,
  });

  sourceFile.addImportDeclaration({
    moduleSpecifier: '@ngneat/elf',
    namedImports: ['Store', 'createState'].map((name) => ({
      kind: StructureKind.ImportSpecifier,
      name,
    })),
  });

  const builders = [
    RequestsCacheBuilder,
    RequestsStatusBuilder,
    ActiveIdBuilder,
    ActiveIdsBuilder,
    EntitiesBuilder,
    PropsBuilder,
    UIEntitiesBuilder,
  ];

  const propsFactories: CallExpression[] = [];

  for (const feature of options.features) {
    for (const builder of builders) {
      if (builder.supports(feature)) {
        const instance = new builder(sourceFile, repoClassDec, options);
        instance.run();
        propsFactories.push(instance.getPropsFactory());
      }
    }
  }

  const state = factory.createCallExpression(
    factory.createIdentifier('createState'),
    undefined,
    propsFactories
  );

  const repoPosition = repoClassDec.getChildIndex();

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    declarations: [
      {
        name: 'store',
        initializer: `new Store({ name: '${camelize(
          storeName
        )}', state, config })`,
      },
    ],
  });

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    declarations: [
      {
        name: '{ state, config }',
        initializer: printNode(state),
      },
    ],
  });

  if (options.template === 'functions') {
    toFunctions(sourceFile, repoClassDec);
  }

  sourceFile.formatText({ indentSize: 2 });

  return sourceFile.getText();
}

function toFunctions(sourceFile: SourceFile, classDec: ClassDeclaration) {
  const exported: string[] = [];

  classDec?.getProperties().forEach((p) => {
    exported.push(`export const ${p.getText()}`);
  });

  classDec?.getMethods().forEach((m) => {
    exported.push(`export function ${m.getText()}`);
  });

  classDec?.remove();

  sourceFile.replaceWithText(
    `${sourceFile.getText()}\n ${exported.join('\n\n')}`
  );
}
