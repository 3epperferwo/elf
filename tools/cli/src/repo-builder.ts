import {
  ClassDeclaration,
  printNode,
  Project,
  QuoteKind,
  SourceFile,
  StructureKind,
  VariableDeclarationKind,
} from 'ts-morph';
import {CallExpression, factory, ScriptTarget} from 'typescript';
import {camelize, capitalize, has} from './utils';
import {Options} from './types';
// @ts-ignore
import * as pluralize from 'pluralize';

export function createRepo(options: Options) {
  const {storeName} = options;

  const project = new Project({
    manipulationSettings: {
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      target: ScriptTarget.ES2015,
    },
  });

  const sourceFile = project.createSourceFile(`repo.ts`, ``);

  sourceFile.addImportDeclaration({
    moduleSpecifier: '@ngneat/elf',
    namedImports: resolveImports(options).map((name) => ({
      kind: StructureKind.ImportSpecifier,
      name,
    })),
  });

  resolveInterfaces(sourceFile, options);

  const state = factory.createCallExpression(
    factory.createIdentifier('createState'),
    undefined,
    resolveFeatures(options)
  );

  sourceFile.addVariableStatement({
    declarationKind: VariableDeclarationKind.Const,
    declarations: [
      {
        name: '{ state, config }',
        initializer: printNode(state),
      },
    ],
  });

  sourceFile.addVariableStatement({
    declarationKind: VariableDeclarationKind.Const,
    declarations: [
      {
        name: 'store',
        initializer: `new Store({ name: '${camelize(
          storeName
        )}', state, config })`,
      },
    ],
  });

  const repo = sourceFile.addClass({
    name: `${capitalize(storeName)}Repository`,
    isExported: true,
  });

  extendRepo(repo, options);

  sourceFile.formatText({indentSize: 2});

  return sourceFile.getText();
}


function extendRepo(repo: ClassDeclaration, options: Options) {
  if (has(options, 'withEntities')) {
    repo.addMember({
      name: `${camelize(options.storeName)}$`,
      kind: StructureKind.Property,
      initializer: `store.pipe(selectAll())`,
    });

    options.crud.forEach((type) => {
      crudOps[type](repo, options);
    });
  }

  if (has(options, 'withActiveId')) {
    extendActiveId(repo, options);
  }

  if (has(options, 'withActiveIds')) {
    extendActiveIds(repo, options);
  }
}


const crudOps: {
  [key in Options['crud'][0]]: (
    repo: ClassDeclaration,
    options: Options
  ) => void;
} = {
  addEntities(repo, options) {
    const singularName = pluralize.singular(options.storeName);

    repo.addMember({
      kind: StructureKind.Method,
      name: `add${capitalize(singularName)}`,
      parameters: [
        {
          name: camelize(singularName),
          type: capitalize(singularName),
        },
      ],
      statements: [`store.reduce(addEntities(${camelize(singularName)}));`],
    });
  },
  updateEntities(repo, options) {
    const singularName = pluralize.singular(options.storeName);

    repo.addMember({
      kind: StructureKind.Method,
      name: `update${capitalize(singularName)}`,
      parameters: [
        {
          name: 'id',
          type: `${capitalize(singularName)}['${options.idKey}']`,
        },
        {
          name: camelize(singularName),
          type: `Partial<${capitalize(singularName)}>`,
        },
      ],
      statements: [
        `store.reduce(updateEntities(id, ${camelize(singularName)}));`,
      ],
    });
  },
  removeEntities(repo, options) {
    const singularName = pluralize.singular(options.storeName);

    repo.addMember({
      kind: StructureKind.Method,
      name: `delete${capitalize(singularName)}`,
      parameters: [
        {
          name: 'id',
          type: `${capitalize(singularName)}['${options.idKey}']`,
        },
      ],
      statements: [`store.reduce(removeEntities(id));`],
    });
  },
};

function extendActiveId(repo: ClassDeclaration, options: Options) {
  const singularName = pluralize.singular(options.storeName);

  repo.addMember({
    kind: StructureKind.Method,
    name: `setActiveId`,
    parameters: [
      {
        name: 'id',
        type: `${capitalize(singularName)}['${options.idKey}']`,
      },
    ],
    statements: [`store.reduce(setActiveId(id));`],
  });

  repo.insertMember(0, {
    name: `active${capitalize(camelize(singularName))}$`,
    kind: StructureKind.Property,
    initializer: `store.pipe(selectActiveEntity())`,
  });
}

function extendActiveIds(repo: ClassDeclaration, options: Options) {
  const singularName = pluralize.singular(options.storeName);

  repo.addMember({
    kind: StructureKind.Method,
    name: `toggleActiveIds`,
    parameters: [
      {
        name: 'ids',
        type: `Array<${capitalize(singularName)}['${options.idKey}']>`,
      },
    ],
    statements: [`store.reduce(setActiveIds(ids));`],
  });

  repo.insertMember(0, {
    name: `active${capitalize(camelize(options.storeName))}$`,
    kind: StructureKind.Property,
    initializer: `store.pipe(selectActiveEntities())`,
  });
}

function resolveImports(options: Options) {
  const base = [
    'Store',
    'createState',
    ...options.features,
    ...(options.crud || []),
  ];

  if (has(options, 'withEntities')) {
    base.push('selectAll');
  }

  if (has(options, 'withActiveId')) {
    base.push('setActiveId', 'selectActiveEntity');
  }

  if (has(options, 'withActiveIds')) {
    base.push('toggleActiveIds', 'selectActiveEntities');
  }

  return base;
}

const features: {
  [key in Options['features'][0]]: (options: Options) => CallExpression;
} = {
  withProps(options) {
    return factory.createCallExpression(
      factory.createIdentifier('withProps'),
      [
        factory.createTypeReferenceNode(
          factory.createIdentifier(`${capitalize(options.storeName)}Props`),
          undefined
        ),
      ],
      [factory.createObjectLiteralExpression([], false)]
    );
  },
  withEntities(options) {
    const type: any[] = [
      factory.createTypeReferenceNode(
        factory.createIdentifier(capitalize(pluralize.singular(options.storeName))),
        undefined
      ),
    ]

    if (options.idKey !== 'id') {
      type.push(factory.createLiteralTypeNode(factory.createStringLiteral("_id", true)))
    }

    let props: any[] = [];

    if (options.idKey !== 'id') {
      props = [factory.createObjectLiteralExpression(
        [factory.createPropertyAssignment(
          factory.createIdentifier("idKey"),
          factory.createStringLiteral("_id", true)
        )],
        false
      )]
    }

    return factory.createCallExpression(
      factory.createIdentifier("withEntities"),
      type,
      props
    )
  },
  withUIEntities(options) {
    const type: any[] = [
      factory.createTypeReferenceNode(
        factory.createIdentifier(`${capitalize(pluralize.singular(options.storeName))}UI`),
        undefined
      ),
    ]

    if (options.idKey !== 'id') {
      type.push(factory.createLiteralTypeNode(factory.createStringLiteral("_id", true)))
    }

    let props: any[] = [];

    if (options.idKey !== 'id') {
      props = [factory.createObjectLiteralExpression(
        [factory.createPropertyAssignment(
          factory.createIdentifier("idKey"),
          factory.createStringLiteral("_id", true)
        )],
        false
      )]
    }

    return factory.createCallExpression(
      factory.createIdentifier("withUIEntities"),
      type,
      props
    )
  },
  withStatus() {
    return factory.createCallExpression(
      factory.createIdentifier('withStatus'),
      undefined,
      []
    );
  },
  withCache() {
    return factory.createCallExpression(
      factory.createIdentifier('withCache'),
      undefined,
      []
    );
  },
  withActiveId() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveId'),
      undefined,
      []
    );
  },
  withActiveIds() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveIds'),
      undefined,
      []
    );
  },
};

function resolveFeatures(options: Options) {
  return options.features.map((f) => features[f](options));
}

function resolveInterfaces(sourceFile: SourceFile, options: Options) {
  const storeName = options.storeName;

  if (has(options, 'withProps')) {
    sourceFile.addInterface({
      name: `${capitalize(storeName)}Props`,
      isExported: true,
    });
  }

  if (has(options, 'withEntities')) {
    sourceFile.addInterface({
      name: capitalize(pluralize.singular(storeName)),
      isExported: true,
      properties: [
        {
          name: options.idKey,
          type: 'number',
        },
      ],
    });
  }

  if (has(options, 'withUIEntities')) {
    sourceFile.addInterface({
      name: `${capitalize(pluralize.singular(storeName))}UI`,
      isExported: true,
      properties: [
        {
          name: options.idKey,
          type: 'number',
        },
      ],
    });
  }
}
