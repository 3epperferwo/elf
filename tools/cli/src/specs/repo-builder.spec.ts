import { createRepo } from '../builders/repo-builder';
import { baseFeatures } from '../types';

describe('Repo Builder', () => {
  it('withProps', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withProps'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withProps');
  });

  it('withCacheStatus', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withCacheStatus'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withCacheStatus');
  });

  it('withStatus', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withStatus'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withStatus');
  });

  it('withActiveId', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveId'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveId');
  });

  it('withActiveIds', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveIds'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveIds');
  });

  it('withEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'removeEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withEntities idKey', () => {
    const output = createRepo({
      idKey: '_id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'removeEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withUIEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withUIEntities'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withUIEntities');
  });

  it('all', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['removeEntities', 'addEntities', 'updateEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('all');
  });
});
