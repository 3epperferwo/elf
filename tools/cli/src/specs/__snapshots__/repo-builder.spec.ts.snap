// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repo Builder all: all 1`] = `
"import { Store, createState, withProps, withEntities, selectAll, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds, withCacheStatus, selectCacheStatus, setCacheStatus, CacheStatus, withStatus, selectStatus, setStatus, StatusState } from '@ngneat/elf';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

export interface TodosProps {
}

const { state, config } = createState(withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withCacheState(), withStatus());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  status$ = store.pipe(selectStatus());
  cacheStatus$ = store.pipe(selectCacheStatus());
  activeTodos$ = store.pipe(selectActiveEntities());
  activeTodo$ = store.pipe(selectActiveEntity());
  todos$ = store.pipe(selectAll());

  deleteTodo(id: Todo['id']) {
    store.reduce(deleteEntities(id));
  }

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  setActiveId(id: Todo['id']) {
    store.reduce(setActiveId(id));
  }

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.reduce(toggleActiveIds(ids));
  }

  updateCacheStatus(value: CacheStatus) {
    store.reduce(setCacheStatus(value));
  }

  updateStatus(value: StatusState) {
    store.reduce(setStatus(value));
  }
}
"
`;

exports[`Repo Builder withActiveId: withActiveId 1`] = `
"import { Store, createState, withActiveId, selectActiveEntity, setActiveId } from '@ngneat/elf';

const { state, config } = createState(withActiveId());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  activeTodo$ = store.pipe(selectActiveEntity());

  setActiveId(id: Todo['id']) {
    store.reduce(setActiveId(id));
  }
}
"
`;

exports[`Repo Builder withActiveIds: withActiveIds 1`] = `
"import { Store, createState, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf';

const { state, config } = createState(withActiveIds());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  activeTodos$ = store.pipe(selectActiveEntities());

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.reduce(toggleActiveIds(ids));
  }
}
"
`;

exports[`Repo Builder withCacheStatus: withCacheStatus 1`] = `
"import { Store, createState, withCacheStatus, selectCacheStatus, setCacheStatus, CacheStatus } from '@ngneat/elf';

const { state, config } = createState(withCacheState());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  cacheStatus$ = store.pipe(selectCacheStatus());

  updateCacheStatus(value: CacheStatus) {
    store.reduce(setCacheStatus(value));
  }
}
"
`;

exports[`Repo Builder withEntities idKey: withEntities 1`] = `
"import { Store, createState, withEntities, selectAll, addEntities, updateEntities, deleteEntities } from '@ngneat/elf';

export interface Todo {
  _id: number;
}

const { state, config } = createState(withEntities<Todo, '_id'>({ idKey: '_id' }));
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  todos$ = store.pipe(selectAll());

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['_id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['_id']) {
    store.reduce(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withEntities: withEntities 1`] = `
"import { Store, createState, withEntities, selectAll, addEntities, updateEntities, deleteEntities } from '@ngneat/elf';

export interface Todo {
  id: number;
}

const { state, config } = createState(withEntities<Todo>());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  todos$ = store.pipe(selectAll());

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['id']) {
    store.reduce(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withProps: withProps 1`] = `
"import { Store, createState, withProps } from '@ngneat/elf';

export interface TodosProps {
}

const { state, config } = createState(withProps<TodosProps>({}));
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
}
"
`;

exports[`Repo Builder withStatus: withStatus 1`] = `
"import { Store, createState, withStatus, selectStatus, setStatus, StatusState } from '@ngneat/elf';

const { state, config } = createState(withStatus());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  status$ = store.pipe(selectStatus());

  updateStatus(value: StatusState) {
    store.reduce(setStatus(value));
  }
}
"
`;

exports[`Repo Builder withUIEntities: withUIEntities 1`] = `
"import { Store, createState, withUIEntities } from '@ngneat/elf';

export interface TodoUI {
  id: number;
}

const { state, config } = createState(withUIEntities<TodoUI>());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
}
"
`;
